Mono's msbuild can't build Blazor WASM projects

## Steps to Reproduce
```
$ dotnet new -i Microsoft.AspNetCore.Components.WebAssembly.Templates::3.2.0-preview3.20168.3
$ dotnet new blazorwasm -o Wasm
$ cd Wasm
$ msbuild Wasm.csproj
```
### Current Behavior

Build fails with:
```
_ResolveBlazorInputs:
  Creating directory "obj/Debug/netstandard2.1/blazor/".
/Users/rodrigo/.nuget/packages/microsoft.aspnetcore.components.webassembly.build/3.2.0-preview3.20168.3/targets/Blazor.MonoRuntime.targets(348,5): error : Specified cast is not valid. [/private/tmp/BlazorWasm/BlazorWasm.csproj]
Done Building Project "/private/tmp/BlazorWasm/BlazorWasm.csproj" (default targets) -- FAILED.
```

### Expected Behavior

Build succeeds

## On which platforms did you notice this

[x] macOS
[ ] Linux
[ ] Windows

**Version Used**:

6.10.0.101

## Stacktrace

See https://github.com/dotnet/aspnetcore/issues/20259 for more information and the MSBuild stack trace. From the last comment from @pranavkm seems the issue is with the DataContractJsonSerializer implementation in Mono

@pranavkm Do you happen to have the `ConfigurationFiles` [here](https://github.com/dotnet/aspnetcore/blob/93eec70bb84fab402bcec49d356fb1e7b65a3464/src/Components/WebAssembly/Build/src/Tasks/GenerateBlazorBootJson.cs#L127)?
When running a sample [script]( https://gist.github.com/mdh1418/4e7c0d4ba38de12b3c834dfad3d9e34e#file-datacontractjsonserializerrepro-cs) with system mono 6.10 and 6.12 `csc`+`mono`, 
I'm getting the error below.
```Unhandled Exception:
System.Configuration.ConfigurationErrorsException: Failed to load configuration section for dataContractSerializer.
  at System.Runtime.Serialization.Configuration.DataContractSerializerSection.UnsafeGetSection () [0x00030] in <e3b7c6553344454d8b9272ffa3726b5a>:0
  at System.Runtime.Serialization.DataContract.get_ConfigSection () [0x00007] in <e3b7c6553344454d8b9272ffa3726b5a>:0
  at System.Runtime.Serialization.DataContract.LoadKnownTypesFromConfig (System.Type type, System.Collections.Generic.Dictionary`2[TKey,TValue] typesChecked, System.Collections.Generic.Dictionary`2[System.Xml.XmlQualifiedName,System.Runtime.Serialization.DataContract]& knownDataContracts) [0x00000] in <e3b7c6553344454d8b9272ffa3726b5a>:0
  at System.Runtime.Serialization.DataContract.ImportKnownTypeAttributes (System.Type type, System.Collections.Generic.Dictionary`2[TKey,TValue] typesChecked, System.Collections.Generic.Dictionary`2[System.Xml.XmlQualifiedName,System.Runtime.Serialization.DataContract]& knownDataContracts) [0x00228] in <e3b7c6553344454d8b9272ffa3726b5a>:0
  at System.Runtime.Serialization.DataContract.ImportKnownTypeAttributes (System.Type type) [0x00008] in <e3b7c6553344454d8b9272ffa3726b5a>:0
  at System.Runtime.Serialization.ClassDataContract+ClassDataContractCriticalHelper.get_KnownDataContracts () [0x00031] in <e3b7c6553344454d8b9272ffa3726b5a>:0
  at System.Runtime.Serialization.ClassDataContract.get_KnownDataContracts () [0x00000] in <e3b7c6553344454d8b9272ffa3726b5a>:0
  at System.Runtime.Serialization.XmlObjectSerializerWriteContext.SerializeWithoutXsiType (System.Runtime.Serialization.DataContract dataContract, System.Runtime.Serialization.XmlWriterDelegator xmlWriter, System.Object obj, System.RuntimeTypeHandle declaredTypeHandle) [0x0002a] in <e3b7c6553344454d8b9272ffa3726b5a>:0
  at System.Runtime.Serialization.Json.DataContractJsonSerializer.InternalWriteObjectContent (System.Runtime.Serialization.XmlWriterDelegator writer, System.Object graph) [0x000a4] in <e3b7c6553344454d8b9272ffa3726b5a>:0
  at System.Runtime.Serialization.Json.DataContractJsonSerializer.InternalWriteObject (System.Runtime.Serialization.XmlWriterDelegator writer, System.Object graph) [0x00008] in <e3b7c6553344454d8b9272ffa3726b5a>:0
  at System.Runtime.Serialization.XmlObjectSerializer.InternalWriteObject (System.Runtime.Serialization.XmlWriterDelegator writer, System.Object graph, System.Runtime.Serialization.DataContractResolver dataContractResolver) [0x00000] in <e3b7c6553344454d8b9272ffa3726b5a>:0
  at System.Runtime.Serialization.XmlObjectSerializer.WriteObjectHandleExceptions (System.Runtime.Serialization.XmlWriterDelegator writer, System.Object graph, System.Runtime.Serialization.DataContractResolver dataContractResolver) [0x00073] in <e3b7c6553344454d8b9272ffa3726b5a>:0
  at System.Runtime.Serialization.XmlObjectSerializer.WriteObjectHandleExceptions (System.Runtime.Serialization.XmlWriterDelegator writer, System.Object graph) [0x00000] in <e3b7c6553344454d8b9272ffa3726b5a>:0
  at System.Runtime.Serialization.Json.DataContractJsonSerializer.WriteObject (System.Xml.XmlDictionaryWriter writer, System.Object graph) [0x00017] in <e3b7c6553344454d8b9272ffa3726b5a>:0
  at monotest2.Program.Main (System.String[] args) [0x00079] in <ca3ddf873cbb4752b53149123e610ab2>:0
[ERROR] FATAL UNHANDLED EXCEPTION: System.Configuration.ConfigurationErrorsException: Failed to load configuration section for dataContractSerializer.
  at System.Runtime.Serialization.Configuration.DataContractSerializerSection.UnsafeGetSection () [0x00030] in <e3b7c6553344454d8b9272ffa3726b5a>:0
  at System.Runtime.Serialization.DataContract.get_ConfigSection () [0x00007] in <e3b7c6553344454d8b9272ffa3726b5a>:0
  at System.Runtime.Serialization.DataContract.LoadKnownTypesFromConfig (System.Type type, System.Collections.Generic.Dictionary`2[TKey,TValue] typesChecked, System.Collections.Generic.Dictionary`2[System.Xml.XmlQualifiedName,System.Runtime.Serialization.DataContract]& knownDataContracts) [0x00000] in <e3b7c6553344454d8b9272ffa3726b5a>:0
  at System.Runtime.Serialization.DataContract.ImportKnownTypeAttributes (System.Type type, System.Collections.Generic.Dictionary`2[TKey,TValue] typesChecked, System.Collections.Generic.Dictionary`2[System.Xml.XmlQualifiedName,System.Runtime.Serialization.DataContract]& knownDataContracts) [0x00228] in <e3b7c6553344454d8b9272ffa3726b5a>:0
  at System.Runtime.Serialization.DataContract.ImportKnownTypeAttributes (System.Type type) [0x00008] in <e3b7c6553344454d8b9272ffa3726b5a>:0
  at System.Runtime.Serialization.ClassDataContract+ClassDataContractCriticalHelper.get_KnownDataContracts () [0x00031] in <e3b7c6553344454d8b9272ffa3726b5a>:0
  at System.Runtime.Serialization.ClassDataContract.get_KnownDataContracts () [0x00000] in <e3b7c6553344454d8b9272ffa3726b5a>:0
  at System.Runtime.Serialization.XmlObjectSerializerWriteContext.SerializeWithoutXsiType (System.Runtime.Serialization.DataContract dataContract, System.Runtime.Serialization.XmlWriterDelegator xmlWriter, System.Object obj, System.RuntimeTypeHandle declaredTypeHandle) [0x0002a] in <e3b7c6553344454d8b9272ffa3726b5a>:0
  at System.Runtime.Serialization.Json.DataContractJsonSerializer.InternalWriteObjectContent (System.Runtime.Serialization.XmlWriterDelegator writer, System.Object graph) [0x000a4] in <e3b7c6553344454d8b9272ffa3726b5a>:0
  at System.Runtime.Serialization.Json.DataContractJsonSerializer.InternalWriteObject (System.Runtime.Serialization.XmlWriterDelegator writer, System.Object graph) [0x00008] in <e3b7c6553344454d8b9272ffa3726b5a>:0
  at System.Runtime.Serialization.XmlObjectSerializer.InternalWriteObject (System.Runtime.Serialization.XmlWriterDelegator writer, System.Object graph, System.Runtime.Serialization.DataContractResolver dataContractResolver) [0x00000] in <e3b7c6553344454d8b9272ffa3726b5a>:0
  at System.Runtime.Serialization.XmlObjectSerializer.WriteObjectHandleExceptions (System.Runtime.Serialization.XmlWriterDelegator writer, System.Object graph, System.Runtime.Serialization.DataContractResolver dataContractResolver) [0x00073] in <e3b7c6553344454d8b9272ffa3726b5a>:0
  at System.Runtime.Serialization.XmlObjectSerializer.WriteObjectHandleExceptions (System.Runtime.Serialization.XmlWriterDelegator writer, System.Object graph) [0x00000] in <e3b7c6553344454d8b9272ffa3726b5a>:0
  at System.Runtime.Serialization.Json.DataContractJsonSerializer.WriteObject (System.Xml.XmlDictionaryWriter writer, System.Object graph) [0x00017] in <e3b7c6553344454d8b9272ffa3726b5a>:0
  at monotest2.Program.Main (System.String[] args) [0x00079] in <ca3ddf873cbb4752b53149123e610ab2>:0
```

Even though I get issues using system mono, when creating a console app and running `dotnet build` + `dotnet run` it seems to work fine [BlazorIssue.zip](https://github.com/mono/mono/files/4462779/BlazorIssue.zip)
```
~/BlazorIssue$dotnet build
Microsoft (R) Build Engine version 16.5.0+d4cbfca49 for .NET Core
Copyright (C) Microsoft Corporation. All rights reserved.

  Restore completed in 40.99 ms for /Users/mdhwang/BlazorIssue/BlazorIssue.csproj.
  BlazorIssue -> /Users/mdhwang/BlazorIssue/bin/Debug/netcoreapp3.1/BlazorIssue.dll

Build succeeded.
    0 Warning(s)
    0 Error(s)

Time Elapsed 00:00:00.92
~/BlazorIssue$dotnet run
Hello World
```

--- a/mcs/class/System.Runtime.Serialization/ReferenceSources/JsonFormatWriterGenerator_static.cs
+++ b/mcs/class/System.Runtime.Serialization/ReferenceSources/JsonFormatWriterGenerator_static.cs
@@ -423,8 +423,10 @@ namespace System.Runtime.Serialization.Json
 			XmlDictionaryString namespaceLocal = null;
 			if (nameLocal != null && nameLocal is string)
 				writer.WriteStartElement ((string) name, null);
-			else
+			else if (name is XmlDictionaryString)
 				writer.WriteStartElement ((XmlDictionaryString) name, null);
+			else
+				writer.WriteStartElement (name.ToString(), null);
 		}
 
 		void WriteEndElement ()

